src/hangar/dataset/common.py

```diff
@@ -0,0 +1,69 @@
+import typing
+from typing import Union, Sequence, List, Set, Tuple
+
+if typing.TYPE_CHECKING:
+    import numpy as np
+    from hangar.columns.column import ModifierTypes as Columns
+
+
+class Dataset:
+    """Dataset class that does the initial checks to verify whether the provided
+    columns can be arranged together as a dataset. These verifications are done on the
+    keys of each column. If ``keys`` argument is ``None``, initializer of this class
+    makes the viable key list by checking the local and common keys across all columns.
+    If ``keys`` argument is provided, then it verifies whether the provided keys are
+    good candidates or not. After all the verification, It provides the ``__getitem__``
+    method to the callee for consumption
+
+
+    Parameters
+    ----------
+    columns : :class:`~hangar.columns.column.Columns` or a Sequence
+        A single column object of a sequence the column objects
+    keys : Sequence[str]
+        An sequence collection of sample names. If given only those samples will
+        fetched from the column
+    """
+
+    def __init__(self, columns: Sequence['Columns'], keys: Sequence[str] = None):
+        if not isinstance(columns, (list, tuple, set)):
+            columns = (columns,)
+        if len(columns) == 0:
+            raise ValueError('len(columns) cannot == 0')
+        all_keys: List[set] = []
+        all_remote_keys: List[set] = []
+        for col in columns:
+            if col.iswriteable is True:
+                raise TypeError(f'Cannot load columns opened in `write-enabled` checkout.')
+            all_keys.append(set(col.keys()))
+            all_remote_keys.append(set(col.remote_reference_keys))
+        common_keys: Set[str] = set.intersection(*all_keys)
+        remote_keys: Set[str] = set.union(*all_remote_keys)
+        common_local_keys: Set[str] = common_keys.difference(remote_keys)
+        if keys:
+            if not isinstance(keys, (list, tuple, set)):
+                raise TypeError('keys must be a list/tuple/set of hangar sample keys')
```

Comment:
I'd get rid of this check; let the user duck type it if they want. As long as it can be turned into a set and/or iterated over, it should be fine.

Reply:
The danger here is if the user passes a string as a single key, our downstream operation converts it into a set of characters (which might not even possess an order) in that string and raises an error saying those samples not found in the column which is confusing

Reply:
I just went through and cleaned up this module a bunch when I rebased ontop of 0.5.

I ended up just doing an `isiterable(keys)` check and that's about it... Even though someone could forget and pass a single sting in (which would be split up as you mentioned sherin), I figured the `KeyError` would suffice. If you passed in an argument `"foosample"` and got an error message saying: `KeyError: samples ['f', 'o', 'o', 's', 'a', 'm', 'p', 'l', 'e'] do not exist ...`, it should be pretty clear what happened? 

Let me know if i'm missing some corner case here?